OpenAQ web API v3

zmoon opened the pull request:
Resolves #198 

Some changes from v2 to v3:
* Have to request data from each sensor separately
  - each sensor has data for one parameter (ID) only
  - to facilitate selecting by country, e.g., have to know all the available locations/sensors, which takes a minute or two to get (implemented caching to help)
* Some differences in the behavior wrt. specifying time period in the request
* No more search radius
* No more sensor type (low-cost vs reference-grade), just the `isMonitor` bool
* hourly- and daily-average data, in addition to raw
* ...

Note: `filelock` should be moved to the optional dependencies section of the Conda env spec and included in the conda-forge recipe.

cc: @jordanschnell

zmoon committed:
Start from a copy of v2

zmoon committed:
To v3, starting with parameters

the returned list is shorter now, and API key seems
to be required now

zmoon committed:
Get locations working again

though not taking all the sensor info

zmoon committed:
Return sensor IDs for location

zmoon committed:
Search radius doesn't seem to be a thing in v3

zmoon committed:
Sensor ID string

zmoon committed:
WIP: get-data

zmoon committed:
Should be singular

zmoon committed:
Commit initial tests

zmoon committed:
WIP: long format data working so far

zmoon committed:
Add metadata

zmoon committed:
snake_case

zmoon committed:
Get sensors for a location (ID)

zmoon committed:
WIP: some time stuff, testing more locs

zmoon committed:
Cache locations JSON data

currently 19790 locations, ~ 27 MB JSON

zmoon committed:
Don't confirm n=0 on next page when get n<limit

seems to help with overall time perf
since fewer changes to be rate-limited

zmoon committed:
Request specific sensors; fix tests

zmoon committed:
Get wide fmt working

zmoon committed:
No time splitting by default

since we are already split a lot, given that we already have to
request each sensor separately

zmoon committed:
Clean up todos

zmoon committed:
Snake case check return bool instead

zmoon committed:
The one sec no impact for this site

but it's also raw sub-hourly data

zmoon committed:
Add options for hourly and daily

was using "raw" before, which seems to often be hourly,
but not always

zmoon committed:
Fix time queries for daily

zmoon committed:
Clean up todo

pretty sure this is handled now

though "raw" data that is actually all on the hour would be missing
the last hour

zmoon committed:
Ensure raw data in requested range

zmoon committed:
Test hourly and daily too in sensor limit case

zmoon committed:
Single time

zmoon committed:
Ignore the cached locations data

zmoon committed:
Compress locations cache file

26 MB -> 1.6 MB

zmoon committed:
Adjust locations cache file locking

to ensure only one worker works to create the locations data cache
(when it is stale or nonexistent)
and the others wait and use it

zmoon committed:
Add imports

zmoon committed:
Avoid warning

zmoon committed:
Tweak location loading message

zmoon committed:
filelock used for openaq-v3

zmoon commented:
Tests failing are ICAP-MME, SSL cert out of date

zmoon commented on monetio/obs/openaq_v3.py:
```diff
@@ -0,0 +1,855 @@
+    if entity is not None:
+        raise NotImplementedError  # TODO: not sure what to use for this
+    if sensor_type is not None:
+        # FIXME: may not be the best approach
+        meta = meta.assign(
+            sensor_type=meta["is_monitor"].map(
+                {
+                    True: "reference grade",
+                    False: "low-cost sensor",
+                }
+            )
+        )
```
@russbiggs is the isMonitor true/false breakdown the same as the old "reference grade" vs "low-cost sensor"?

russbiggs replied:
In v3 it's `monitor`. So `?monitor=true` will return just reference monitors, `?monitor=false` will return just air sensors (aka low-cost sensor). Not including this parameter will include both types.

zmoon replied:
Thanks, seems like my approach here is correct then.

zmoon replied:
`is_monitor` that I'm using in this code block is `isMonitor` from the locations endpoint data. I assume the `monitor` query parameter you mention is consistent with that.

zmoon commented on monetio/obs/openaq_v3.py:
```diff
@@ -0,0 +1,855 @@
+                logger.info(f"request timed out (try {tries}/{retry})")
+                time.sleep(tries + 0.1 * rand())
+            elif r.status_code == 429:
+                # Note: response headers don't seem to include Retry-After
+                # Just `{'detail': 'To many requests'}` in `.json()`
+                logger.info(f"rate limited (try {tries}/{retry})")
+                time.sleep(tries * 5 + 0.2 * rand())
```
@russbiggs do you have any comments about this approach? Would you recommend backing off more initially when rate-limited? Is there some way to get Retry-After? Or is it always the same?

russbiggs replied:
The preferred approach is to use the providers rate limit headers in the HTTP response which you can read about here:

https://docs.openaq.org/using-the-api/rate-limits#rate-limit-headers

In the official OpenAQ python SDK you can see how i'm exposing them, (I'm using httpx but its pretty close to requests).

https://github.com/openaq/openaq-python/blob/main/openaq/shared/responses.py#L147

If you bundle these values with the response (or they would already be there if using the response object directly) then you can access and throttle before the next request something like:

```py
if req.headers["x-ratelimit-remaining"] < 1:
    time.sleep(req.headers["x-ratelimit-reset"])
# else proceed to next request
```

Then you will sleep for exactly long as is needed until the rate limit resets and you'll get a fresh allocation of requests, no need to exponential back off or guessing.

zmoon replied:
Gotcha, I'll add that possible wait for the 200 case, I can see those x-ratelimit response headers are there. However, for the 429 case, which can still happen with multi-threaded requesting, it seems those headers are not currently included.

russbiggs replied:
Good catch, if you submit a ticket here https://github.com/openaq/openaq-api-v2 i dont see why they couldnt be added. i think it would be a relatively straightforward add

zmoon committed:
Use the x-ratelimit headers

currently not available in the 429 case though
but 60 should be the max

with this, should never get rate limited if using one thread only

zmoon committed:
Lower default wait to 30 s

zmoon committed:
Stop multi-threaded execution on error

note only tasks that weren't started can be cancelled, so there
is still lag between the time of first exception and when it stops

but it will even stop (after some time) for ctrl-c now

zmoon committed:
isMobile does indeed map to sensor type this way

got confirmation from OpenAQ

zmoon committed:
Add rate limit info link

zmoon committed:
Fix API key doc link
